### PR TITLE
Fix slow tests sometimes failing

### PR DESCRIPTION
**Pull request by flodiebold**
In some situations we reloaded the workspace in the tests after having reported
to be ready. There's two fixes here:
1. Add a version to the VFS config and include that version in progress reports,
so that we don't think we're done prematurely;
2. Delay status transitions until after changes are applied. Otherwise the last
change during loading can potentially trigger a workspace reload, if it contains
interesting changes.

### CODE REVIEW

File: crates/rust-analyzer/src/reload.rs [hunk 1]
```diff
@@ -50,6 +50,15 @@ impl GlobalState {
...
             Status::Ready { .. } | Status::Invalid => (),
         }
+        log::info!(
+            "Reloading workspace because of the following changes: {}",
+            changes
+                .iter()
+                .filter(|(path, kind)| is_interesting(path, *kind))
+                .map(|(path, kind)| format!("{}/{:?}", path.display(), kind))
+                .collect::<Vec<_>>()
+                .join(", ")
```
Comment (matklad) [inline crates/rust-analyzer/src/reload.rs:60]: I think we have itertools here

File: crates/vfs/src/loader.rs [hunk 2]
```diff
@@ -32,6 +32,9 @@ pub struct Directories {
 /// [`Handle`]'s configuration.
 #[derive(Debug)]
 pub struct Config {
+    /// Version number to associate progress updates to the right config
+    /// version.
+    pub version: usize,
```
Comment (matklad) [inline crates/vfs/src/loader.rs:37]:
```suggestion
    pub version: u32,
```

I can't articulate why, but I prefer to use `u32` as goto "an integer type". It feels proper not to depend on pointer width by default.

File: crates/rust-analyzer/src/global_state.rs [hunk 3]
```diff
@@ -67,6 +67,7 @@ pub(crate) struct GlobalState {
     req_queue: ReqQueue,
     pub(crate) task_pool: Handle<TaskPool<Task>, Receiver<Task>>,
     pub(crate) loader: Handle<Box<dyn vfs::loader::Handle>, Receiver<vfs::loader::Message>>,
+    pub(crate) vfs_config_version: usize,
```
Comment (matklad) [inline crates/rust-analyzer/src/global_state.rs:70]: I wonder if this bit of state belongs to `vfs::Handle`? That way, the client won't be able to mess-up version numbers. That is,

```
pub trait Handle: fmt::Debug {
    fn set_config(&mut self, config: Config) -> u32
```

Or maybe even `-> OpaqueConfigId`, if we go full proper software engineering.

**Review by matklad — COMMENTED** [review]
lgtm, but maybe it makes sense to move responsibilities for managing the version around.

File: crates/rust-analyzer/src/global_state.rs (hunk 3 above)
Comment (flodiebold) [reply crates/rust-analyzer/src/global_state.rs:70]: Hmm, the problem is that `Config` contains the version already, so set_config could only return the version from that...

Comment (matklad) [reply crates/rust-analyzer/src/global_state.rs]: But does it need to contain a version? Seems like the client can set config without a version, and vfs can reply "ok, this is the ID for your config". As in, config is just data, it get's identity only when submitted to vfs for processing.

Comment (flodiebold) [reply crates/rust-analyzer/src/global_state.rs]: We'd still need this field to keep track of the config version though.

Comment (matklad) [reply crates/rust-analyzer/src/global_state.rs]: Yeah, that's true. Still feels more reliable if the vfs owns the incrementing counter, but, indeed, this doesn't make the code actually simpler, so lgtm as is!

**Pull request merged**